lib/http/HttpClient_WinInet.cpp

```diff
@@ -136,6 +139,105 @@ class WinInetRequestWrapper
+            &data[0], &stringSize);
+        if (getStringResult != ERROR_SUCCESS) {
+            return "";
+        }
+
+        // Remove the null terminator from Win32 API. The std::string will already cover this.
+        data.resize(stringSize - 1);
+
+        return data;
+    }
+
+    bool ProxyAuthRequired() {
+        return GetDWORDRegKey(HKEY_CURRENT_USER, kProxyRegKeyPath, "ProxyEnable");
+    }
+
+    void SetProxyCredentials() {
+        // get proxy server from registry
+        std::string proxyServer = GetStringRegKey(HKEY_CURRENT_USER, kProxyRegKeyPath, "ProxyServer");
+        if (proxyServer.empty()) {
+            return;
+        }
+
+        std::wstring formattedProxyServer;
+
+        // If there is only one colon, then it is most likely of form IP:PORT and we just need
+        // to remove the port. If there are two colons, then it is most likely of form SCHEME:IP:PORT
+        int colonCount = 0;
+        for (const char& c : proxyServer) {
+            if (c == ':') {
+                colonCount++;
+            }
+        }
+
+        if (colonCount == 1) {
+            const auto colonPos = proxyServer.find_first_of(':');
+            auto portRemoved = proxyServer.substr(0, colonPos);
+            formattedProxyServer = std::wstring(portRemoved.begin(), portRemoved.end());
+        } else if (colonCount == 2) {
+            const auto firstColonPos = proxyServer.find_first_of(':');
+            auto schemeRemoved = proxyServer.substr(firstColonPos + 3);
+
+            const auto lastColonPos = schemeRemoved.find_last_of(':');
+            auto portAndSchemeRemoved = schemeRemoved.substr(0, lastColonPos);
+
+            formattedProxyServer = std::wstring(portAndSchemeRemoved.begin(), 
+                                                portAndSchemeRemoved.end());
+        } else {
+            formattedProxyServer = std::wstring(proxyServer.begin(), 
+                                                proxyServer.end());
+        }
+
+        // get proxy credentials from credential manager
+        PCREDENTIALW cred;
+        if (!::CredReadW(formattedProxyServer.c_str(), CRED_TYPE_GENERIC, 0, &cred)) {
+            return;
+        }
+        wchar_t* proxyUser = cred->UserName;
+        wchar_t* proxyPass = (wchar_t*)cred->CredentialBlob;
+
+        // set proxy credentials
+        if (proxyUser) {
+            ::InternetSetOptionW(m_hWinInetRequest,
+                                 INTERNET_OPTION_PROXY_USERNAME,
+                                 static_cast<void*>(proxyUser),
+                                 static_cast<DWORD>(wcslen(proxyUser) + 1));
+        }
+
+        if (proxyPass) {
+            ::InternetSetOptionW(m_hWinInetRequest,
+                                 INTERNET_OPTION_PROXY_PASSWORD,
+                                 static_cast<void*>(proxyPass),
+                                 static_cast<DWORD>(wcslen(proxyPass) + 1));
+        }
+
+        ::CredFree(cred);
```

Comment:
You should either wrap PCREDENTIALW in a helper type whose destructor calls ::CredFree, or refactor 216-238 into a helper function which is marked `noexcept`. That way if at any point (including if someone adds new code which can raise exceptions) an exception is raised there's no chance cred is leaked.

Reply:
The former is preferred as it's more resilient.
```c++
struct CredHolder final
{
    PCREDENTIALW cred;
    ~CredHolder() noexcept
    {
        ::CredFree(cred);
    }
};
```